python-package/lightgbm/engine.py

```diff
@@ -125,6 +127,18 @@ def train(
     """
     # create predictor first
     params = copy.deepcopy(params)
+    eval_metric = params.get('metric')
+    # merge metrics through params and feval
+    if eval_metric is not None:
+        eval_metric_list = copy.deepcopy(eval_metric)
+        if not isinstance(eval_metric, list):
+            eval_metric_list = [eval_metric_list] if isinstance(eval_metric_list, str) else list(eval_metric_list)
```

Comment:
When eval_metric is a single callable object, (e.g., when use a callable function for `metric` in parameter), this will result in an error `TypeError: 'function' object is not iterable` since we are converting a function to a list.

Reply:
I see here three cases to handle:

1. `eval_metric_list` is str -> `[eval_metric_list]`
2. `eval_metric_list` is callable -> `[eval_metric_list]`
3.  `eval_metric_list` is a set -> `list(eval_metric_list)`

in case `eval_metric_list` is a tuple then `list(eval_metric_list)` works.